scanpipe/pipes/d2d.py

```diff
@@ -524,3 +576,97 @@ def match_purldb(project, extensions, logger=None):
             start_time=start_time,
         )
         _match_purldb_resource(project, to_resource)
+
+
+def map_javascript(project, logger=None):
+    """Map a packed or minified JavaScript, TypeScript, CSS and SCSS to its source."""
+    project_files = project.codebaseresources.files().only("path")
+
+    from_resources = project_files.from_codebase().filter(
+        extension__in=PROSPECTIVE_JAVASCRIPT_MAP.keys()
+    )
+    to_resources = project_files.to_codebase()
+    resource_count = from_resources.count()
+
+    # For most case in JavaScript we have one-to-many relation (single source file
+    # may generate up to 7-8 different files), in such case if we start by `to`
+    # side we will need to supply another layer of logic to selectively get those
+    # specific 7-8 file that might relate to a particular file from `from` side.
+
+    if logger:
+        logger(
+            f"Mapping {resource_count:,d} from/ resources using javascript map "
+            f"against to/ codebase"
+        )
+
+    resource_iterator = from_resources.iterator(chunk_size=2000)
+    last_percent = 0
+    start_time = timer()
+    for resource_index, from_resource in enumerate(resource_iterator):
+        last_percent = pipes.log_progress(
+            logger,
+            resource_index,
+            resource_count,
+            last_percent,
+            increment_percent=10,
+            start_time=start_time,
+        )
+        _map_javascript_resource(from_resource, to_resources)
+
+
+def _map_javascript_resource(from_resource, to_resources):
+    path = Path(from_resource.path.lstrip("/"))
+    # Skip any dot files.
+    if path.name.startswith("."):
+        return
+
+    basename, from_extension = js.get_basename_and_extension(path.name)
+    path_parts = (path.parent / basename).parts
+    path_parts_len = len(path_parts)
+
+    for index in range(1, path_parts_len - 1):
+        current_parts = path_parts[index:]
+        current_path = "/".join(current_parts)
+
+        any_map = False
+
+        for candidate in PROSPECTIVE_JAVASCRIPT_MAP.get(from_extension, []):
+            query = Q(
+                *[
+                    Q(path__endswith=f"{current_path}{extension}")
+                    for extension in candidate["to_related"]
+                ],
+                _connector=Q.OR,
+            )
```

Comment:
This QuerySet is expensive as it triggers a table scan. We should not call it multiple times within 2 for loops, but rather try to get everything at once and then iter on the values.

@pombredanne We likely want to use the new indexing system here as well, to be generated in `map_javascript` with all the candidates for mapping.